### ISSUE TITLE
[gatsby-plugin-netlify-cms] Preview stylesheet is not registering

### ISSUE THREAD

**Issue by vai0** [gatsbyjs/gatsby#15126]
## Description

After updating `gatsby-plugin-netlify-cms` to `4.1.0` from `3.0.9`, none of the netlify preview templates have styles applied.

What's interesting is if I import `netlify-cms` instead of `netlify-cms-app`, this problem goes away. Looking at the chrome's network tab, it appears `cms.css` is not getting picked up when importing `netlify-cms-app`

cms.js
```javascript
import CMS from `netlify-cms`;
...
CMS.registerPreviewStyle(`cms.css`);
```

It seems like the only [documented difference](https://www.npmjs.com/package/netlify-cms-app) that probably matters between `netlify-cms` and `netlify-cms-app` is `react` and `react-dom` are now peerDependencies - although this doesn't really explain why this is happening. 

### Environment

Run `gatsby info --clipboard` in your project directory and paste the output here.
```
System:
    OS: Linux 4.19 Debian GNU/Linux 9 (stretch) 9 (stretch)
    CPU: (8) x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
    Shell: 4.4.23 -
/nix/store/mcyvp1s45scjkkhyb1m16aqlsb8wr6hs-bash-interactive-4.4-p23/bin/bash
  Binaries:
    Node: 10.15.3 -
/nix/store/47wbs3fgygg6i7za4knbcqpmrwlckqy4-bifrost-env/bin/node
    Yarn: 1.16.0 - /usr/bin/yarn
    npm: 6.4.1 - /nix/store/47wbs3fgygg6i7za4knbcqpmrwlckqy4-bifrost-env/bin/npm
  Languages:
    Python: 3.6.8 - /home/jchi/Projects/bifrost/backend/venv/bin/python
  Browsers:
    Chrome: 73.0.3683.103
    Firefox: 60.7.1
  npmPackages:
    gatsby: 2.10.4 => 2.10.4
    gatsby-cli: 2.7.2 => 2.7.2
    gatsby-link: 2.2.0 => 2.2.0
    gatsby-plugin-intercom-spa: 0.1.0 => 0.1.0
    gatsby-plugin-netlify-cms: 4.1.0-dev-1561486008141 => 4.1.0-dev-1561486008141
    gatsby-plugin-polyfill-io: 1.1.0 => 1.1.0
    gatsby-plugin-react-helmet: 3.1.0 => 3.1.0
    gatsby-plugin-root-import: 2.0.5 => 2.0.5
    gatsby-plugin-sass: 2.1.0 => 2.1.0
    gatsby-plugin-sentry: 1.0.1 => 1.0.1
    gatsby-plugin-sitemap: 2.2.0 => 2.2.0
    gatsby-source-apiserver: 2.1.3 => 2.1.3
    gatsby-source-filesystem: 2.1.0 => 2.1.0
    gatsby-transformer-json: 2.2.0 => 2.2.0
```

**Comment by sidharthachatterjee** [comment]
CC @erquhart

**Comment by vai0** [comment]
keep open - will eventually get a repo up. very busy atm :/

**Comment by vai0** [comment]
As an update - just realized that this issue only exists in development. Post-build the previews render just fine. Took a quick look at the networking tab again and realized that `cms.css` IS being loaded in, but the contents isn't CSS

`cms.css` in development
```

**Comment by erquhart** [comment]
Note that you're registering a URL - Netlify CMS injects a `link` with it's `href` set to `cms.css` into the head of the preview pane iframe. It sounds like there's no file named `cms.css` available during development, probably because Gatsby is adding a hash after the filename so it's `cms-a6fe32.css` or something.

That said, `gatsby-plugin-netlify-cms` is specifically designed so that you don't need to register styles, it automatically injects all of your site styles into the preview pane iframe. Is that not happening for you? And if not, can you help me reproduce?

**Comment by vai0** [comment]
Ah gotcha. Unfortunately our repo is private.  But my [forked version](https://github.com/vai0/gatsby-plugin-netlify-cms/blob/master/src/gatsby-node.js) of `gatsby-plugin-netlify-cms` is public, and I'm free to share files I think may be relevant here... .e.g, here's our custom cms.js. Let me know what else I can do to help.

```
import CMS from "netlify-cms";

import "typeface-lato";
import "typeface-inconsolata";
import "scss/global.scss";

import StandardLandingPagePreview from "cms/preview-templates/StandardLandingPagePreview";
import StandardThankYouPagePreview from "cms/preview-templates/StandardThankYouPagePreview";
import BetaThankYouPagePreview from "cms/preview-templates/BetaThankYouPagePreview";
import EbookLandingPagePreview from "cms/preview-templates/EbookLandingPagePreview";
import WebinarSeriesLandingPagePreview from "cms/preview-templates/WebinarSeriesLandingPagePreview";
import EventMainLandingPagePreview from "cms/preview-templates/EventMainLandingPagePreview";
import EventHappyHourLandingPagePreview from "cms/preview-templates/EventHappyHourLandingPagePreview";
import EventActivityLandingPagePreview from "cms/preview-templates/EventActivityLandingPagePreview";
import EventRoadshowLandingPagePreview from "cms/preview-templates/EventRoadshowLandingPagePreview";

// The following window and global config settings below were taken from here.
// https://github.com/gatsbyjs/gatsby/blob/master/docs/docs/visual-testing-with-storybook.md
// They're required because the netlify-cms runs on a separate webpack config,
// and outside of Gatsby. This ensures any Gatsby components imported into the
// CMS works without errors

// highlight-start
// Gatsby's Link overrides:
// Gatsby defines a global called ___loader to prevent its method calls from creating console errors you override it here
global.___loader = {
    enqueue: () => {},
    hovering: () => {},
};

// Gatsby internal mocking to prevent unnecessary errors
global.__PATH_PREFIX__ = "";

// This is to utilized to override the window.___navigate method Gatsby defines and uses to report what path a Link would be taking us to
window.___navigate = pathname => {
    alert(`This would navigate to: https://www.memsql.com${pathname}`);
};

CMS.registerPreviewTemplate("lp-standard", StandardLandingPagePreview);
CMS.registerPreviewTemplate("lp-ebook", EbookLandingPagePreview);
CMS.registerPreviewTemplate(
    "lp-webinar-series",
    WebinarSeriesLandingPagePreview
);
CMS.registerPreviewTemplate("content", StandardLandingPagePreview);
CMS.registerPreviewTemplate("content-syndication", StandardLandingPagePreview);
CMS.registerPreviewTemplate("programmatic", StandardLandingPagePreview);
CMS.registerPreviewTemplate("sponsored-webcast", StandardLandingPagePreview);
CMS.registerPreviewTemplate("webcast", StandardLandingPagePreview);
CMS.registerPreviewTemplate("web-forms", StandardLandingPagePreview);
CMS.registerPreviewTemplate("other", StandardLandingPagePreview);
CMS.registerPreviewTemplate("lp-thank-you", StandardThankYouPagePreview);
CMS.registerPreviewTemplate("lp-beta-thank-you", BetaThankYouPagePreview);
CMS.registerPreviewTemplate("lp-event-main", EventMainLandingPagePreview);
CMS.registerPreviewTemplate(
    "lp-event-happy-hour",
    EventHappyHourLandingPagePreview
);
CMS.registerPreviewTemplate(
    "lp-event-activity",
    EventActivityLandingPagePreview
);
CMS.registerPreviewTemplate(
    "lp-event-roadshow",
    EventRoadshowLandingPagePreview
);
```

**Comment by erquhart** [comment]
So `gatsby-plugin-netlify-cms` is using this file as an entry point in a dedicated webpack build, and all css/scss files imported through this entry point get injected to your preview pane. Is that not happening? If there are missing styles, it may just be a matter of importing those styles here.

**Comment by vai0** [comment]
The styles are not being imported in development, but they are just fine post-build.

Can you clarify what the difference is between the cms.js in the `gatsby-plugin-netlify-cms` and our custom `cms.js`? If we have our own `cms.js`, does it simply override the existing `gatsby-plugin-netlify-cms`'s `cms.js`?

**Comment by erquhart** [comment]
Yep, your module overrides the default one.

Just checked this out locally and looks like our plugin is busted - its injecting styles direct to the CMS itself, and also seeing the 404 for some styles too. I'll look into this more in the am.

**Comment by vai0** [comment]
@erquhart hm... not sure why yet. if a diff is done between v4.1.6 and the fork - the differences are the additional babel loaders (seemed like the built-in gatsby preset babel loader wasn't reading React.Fragment syntax?), the resolvePaths option, and `netlify-cms` instead of `netlify-cms-app`.

**Comment by vai0** [comment]
As another update, not sure if this helps at all - but I've noticed it accesses the `cms.css` stylesheet at `localhost:9000/admin/cms.css`, which is redirecting me to a gatsby development 404 page.

Whereas after building gatsby - I do indeed see `cms.css` being available at the `/admin/` dir, which explains why the styles work fine post-build.

So it does seem like in development, `cms.css` is not being generated at all. I'm not familiar enough with webpack to determine which configuration causing this only dev. Any ideas? Here's the forked plugin's gatsby-node.js
```
```

**Comment by vai0** [comment]
@erquhart 

I was able to replicate this issue in a fresh repo: https://github.com/vai0/g-netlify-cms-test

What I noticed was that the issue only occurs in `gatsby develop` before `gatsby build` is run, or when the `.cache/ and public/` are removed and `gatsby develop` is run with a clean slate. 

It appears `gatsby develop` isn't creating and referencing it's own version of the `cms.css` - which explains why my previews only worked post-build, and not in development.

Here are the I went through to replicate the issue:

1. Clone the [repo](https://github.com/vai0/g-netlify-cms-test)
2. Run `gatsby develop` and visit http://localhost:8000/admin/#/collections/test_template/entries/sample

    No styles are applied
    ![dev-before-build](https://user-images.githubusercontent.com/1195698/62023030-d3937800-b183-11e9-9cb5-e9f787ca87aa.png)


3. Stop the dev server, run `gatsby build`, then run `gatsby server`, then visit the statically rendered dashboard: http://localhost:9000/admin/#/collections/test_template/entries/sample

    Styles are picked up and applied
    ![build](https://user-images.githubusercontent.com/1195698/62023066-03428000-b184-11e9-8301-15fc37ea0fa0.png)

4. Stop the server, then run `gatsby develop`, and visit the dev dashboard at visit http://localhost:8000/admin/#/collections/test_template/entries/sample

    Styles are now being picked up in development
    ![dev-after-build](https://user-images.githubusercontent.com/1195698/62023091-2bca7a00-b184-11e9-9a77-53c951270e93.png)

5. To prove that the dev server is picking up the cms.css statically built from `gatsby build`, run `gatsby clean`, then rerun `gatsby develop`, and refresh the dashboard again - and as we can see, it's back to being unstyled.

![dev-before-build](https://user-images.githubusercontent.com/1195698/62023139-6a603480-b184-11e9-9740-cee910165525.png)

Here's the `cms.css` response content when the previews are unstyled:
```
<!DOCTYPE html><html><head><meta charSet="utf-8"/><meta http-equiv="x-ua-compatible" content="ie=edge"/><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"/><title data-react-helmet="true"></title><link rel="shortcut icon" href="/icons/icon-48x48.png?v=008654519ce705ac7bc44303a9014606"/><link rel="manifest" href="/manifest.webmanifest"/><meta name="theme-color" content="#663399"/><link rel="apple-touch-icon" sizes="48x48" href="/icons/icon-48x48.png?v=008654519ce705ac7bc44303a9014606"/><link rel="apple-touch-icon" sizes="72x72" href="/icons/icon-72x72.png?v=008654519ce705ac7bc44303a9014606"/><link rel="apple-touch-icon" sizes="96x96" href="/icons/icon-96x96.png?v=008654519ce705ac7bc44303a9014606"/><link rel="apple-touch-icon" sizes="144x144" href="/icons/icon-144x144.png?v=008654519ce705ac7bc44303a9014606"/><link rel="apple-touch-icon" sizes="192x192" href="/icons/icon-192x192.png?v=008654519ce705ac7bc44303a9014606"/><link rel="apple-touch-icon" sizes="256x256" href="/icons/icon-256x256.png?v=008654519ce705ac7bc44303a9014606"/><link rel="apple-touch-icon" sizes="384x384" href="/icons/icon-384x384.png?v=008654519ce705ac7bc44303a9014606"/><link rel="apple-touch-icon" sizes="512x512" href="/icons/icon-512x512.png?v=008654519ce705ac7bc44303a9014606"/><script src="/socket.io/socket.io.js"></script></head><body><noscript id="gatsby-noscript">This app works best with JavaScript enabled.</noscript><div id="___gatsby"></div><script src="/commons.js"></script></body></html>
```
and when it is styled:
```
.test-template{font-family:sans-serif;max-width:1000px;margin:0 auto;padding:100px}.test-template .title{font-size:48px;font-weight:700;color:#00f;margin-bottom:40px}.test-template .subtitle{font-size:24px;font-weight:700;color:green;margin-bottom:40px}.test-template .summary{font-size:16px;color:#639}
```

**Comment by vai0** [comment]
@erquhart @sidharthachatterjee have a reproducable repo up ^. Let me know what I can do to further help debug

**Comment by vai0** [comment]
Can we label as not not stale?

**Comment by erquhart** [comment]
Hey @vai0, haven't forgotten about this.

@erezrokah this will probably be worth considering while working on #18245, as both may be helped by changing how we work with Webpack from the plugin.

**Comment by erezrokah** [comment]
I did some digging. The `css` file is not extracted for development builds.

There are two current issues with the plugin affecting it:
1. The code here: https://github.com/erezrokah/gatsby/blob/0b96c79bf894fb604d66ce8240298e38f55fff1b/packages/gatsby-plugin-netlify-cms/src/gatsby-node.js#L55 doesn't handle rules in the following format:

```
rules: [
        {
          test: /\.less$/,
          use: [
            loaders.miniCssExtract(),
            loaders.css({ importLoaders: 1 }),
            loaders.postcss(),
            `less-loader`,
          ],
        },
      ],
```
2. The `MiniCssExtractPlugin` should not be used in development builds: https://github.com/webpack-contrib/mini-css-extract-plugin/tree/50434b55c9df9cd3c26f2243e23df83e2964c720#advanced-configuration-example. I figured this out after fixing issue number 1 and getting compilation errors while running `gatsby develop`

**Comment by entr** [comment]
@erezrokah  I had a look at your solution. It walks injected link tags and registers them as preview styles but they are still left in parent context and can easily collide with styles of the cms itself.

I tried using the `MiniCssExtractPlugin.loader` loader in development too and it works just fine with my setup using the `{ hmr: true }` option.

**Comment by erezrokah** [comment]
Thanks @entr, let me have another look then

**Comment by erezrokah** [comment]
@entr can you share your setup as a comment to [this issue](https://github.com/gatsbyjs/gatsby/issues/20428)? setting `{ hmr: true }` in the plugin is not that straightforward as it requires modifying existing Gatsby rules (and user configured rules).

**Comment by entr** [comment]
@erezrokah I did a PR #20589

Let me know what you think. Thanks

**Comment by kylesloper** [comment]
Hi, I'm experiencing the same problems as you are except I use eleventy as my ssg. Could you summarize how you fixed this?
